Keyword classifier (#3123)
RasaHQ/rasa: docs/nlu/components.rst

Intent Classifiers
------------------

.. _keyword_intent_classifier:

KeywordIntentClassifier
~~~~~~~~~~~~~~~~~~~~~~~

> :Short: Simple keyword matching intent classifier.

Reviewer: I'd maybe still leave a note saying something along the lines of that it's not sensible to use longterm

Author: 👍

:Outputs: ``intent``

> :Requires: nothing

Reviewer: this is no longer true

Author: What do you actually mean by this? It requires training data only, that's not listed as required by any other classifiers.

:Output-Example:

    .. code-block:: json

        {
            "intent": {"name": "greet", "confidence": 1.0}
        }

:Description:
    This classifier works by searching a message for user-defined keywords.
    The matching is case insensitive and search only for exact matches of the keyword-string.

>     This is intended to augment the other classifiers. In order to use it it must be

Reviewer: I wouldn't say it's necesarily intended to augment other classifiers? Where did this come from?
And why is it dependent on other classifiers?

Author: I should have said something like "_only_ intended to augment others", as in this should not be the only classifier being used.
As far as I know there is nothing in the NLU that compares different classifiers, (is there?). Usually the pipeline has exactly one classifier. In that case the order is important because intent is set by the classifier changing a property of the message object. The order is which multiple classifiers do this is the order in which they are specified in the pipeline, so the keyword classifier must be last or it will have its intent overwritten.

              starspace algorithm in the case ``mu_neg = mu_pos`` and ``use_max_sim_neg = False``.
              See `starspace paper <https://arxiv.org/abs/1709.03856>`_ for details.

.. _keyword_intent_classifier:

KeywordIntentClassifier
~~~~~~~~~~~~~~~~~~~~~~~

:Short: Simple keyword matching intent classifier, intended for small, short-term projects.
:Outputs: ``intent``
:Requires: nothing
:Output-Example:

    .. code-block:: json

        {
            "intent": {"name": "greet", "confidence": 1.0}
        }

:Description:
    This classifier works by searching a message for keywords.

>     The matching is case insensitive and search only for exact matches of the keyword-string.

Reviewer 2: let's leave case sensitivity to tokenizer

Author: Yup, removed the conversion to lowercase in the code and updated the component description in the doc.

              starspace algorithm in the case ``mu_neg = mu_pos`` and ``use_max_sim_neg = False``.
              See `starspace paper <https://arxiv.org/abs/1709.03856>`_ for details.

.. _keyword_intent_classifier:

KeywordIntentClassifier
~~~~~~~~~~~~~~~~~~~~~~~

:Short: Simple keyword matching intent classifier, intended for small, short-term projects.
:Outputs: ``intent``
:Requires: nothing
:Output-Example:

    .. code-block:: json

        {
            "intent": {"name": "greet", "confidence": 1.0}
        }

:Description:
    This classifier works by searching a message for keywords.
    The matching is case sensitive by default and searches only for exact matches of the keyword-string in the user message.
    The keywords for an intent are the examples of that intent in the NLU training data.
    This means the entire example is the keyword, not the individual words in the example.

>     .. note:: This classifier is intended only for small projects or to get started. If you have few training data you can use the ``pretrained_embeddings_spacy`` pipeline to get started.

Reviewer 2: ```suggestion
    .. note:: This classifier is intended only for small projects or to get started. If you have NLU training data you can use one of our pipelines :ref:`choosing-a-pipeline`.
```

Reviewer 2: Please break long lines

Author: I thought lines are done automatically, I'll fix it

Reviewer 2: not for docs

              starspace algorithm in the case ``mu_neg = mu_pos`` and ``use_max_sim_neg = False``.
              See `starspace paper <https://arxiv.org/abs/1709.03856>`_ for details.

.. _keyword_intent_classifier:

KeywordIntentClassifier
~~~~~~~~~~~~~~~~~~~~~~~

:Short: Simple keyword matching intent classifier, intended for small, short-term projects.
:Outputs: ``intent``
:Requires: nothing
:Output-Example:

    .. code-block:: json

        {
            "intent": {"name": "greet", "confidence": 1.0}
        }

:Description:
    This classifier works by searching a message for keywords.
    The matching is case sensitive by default and searches only for exact matches of the keyword-string in the user message.
    The keywords for an intent are the examples of that intent in the NLU training data.
    This means the entire example is the keyword, not the individual words in the example.
    In the case of multiple keywords being found the message will not be classified.

>     .. note:: This classifier is intended only for small projects or to get started.

Reviewer 3: please also suggest what the user should do instead

              starspace algorithm in the case ``mu_neg = mu_pos`` and ``use_max_sim_neg = False``.
              See `starspace paper <https://arxiv.org/abs/1709.03856>`_ for details.

.. _keyword_intent_classifier:

KeywordIntentClassifier
~~~~~~~~~~~~~~~~~~~~~~~

:Short: Simple keyword matching intent classifier, intended for small, short-term projects.
:Outputs: ``intent``
:Requires: nothing
:Output-Example:

    .. code-block:: json

        {
            "intent": {"name": "greet", "confidence": 1.0}
        }

:Description:
    This classifier works by searching a message for keywords.
    The matching is case sensitive by default and searches only for exact matches of the keyword-string in the user message.
    The keywords for an intent are the examples of that intent in the NLU training data.
    This means the entire example is the keyword, not the individual words in the example.

>     In the case of multiple keywords being found the message will classified as having the intent specified

Reviewer 2: with not having the intent?
